src/crypto/DeviceList.js

```diff
@@ -306,41 +306,57 @@ export default class DeviceList {
         ).then((res) => {
             const dk = res.device_keys || {};
 
+            // do each user in a separate promise, to avoid wedging the CPU
+            // (https://github.com/vector-im/riot-web/issues/3158)
+            //
+            // of course we ought to do this in a web worker or similar, but
+            // this serves as an easy solution for now.
+            let prom = q();
             for (const userId of downloadUsers) {
-                console.log('got keys for ' + userId + ':', dk[userId]);
-
-                // map from deviceid -> deviceinfo for this user
-                const userStore = {};
-                const devs = this._sessionStore.getEndToEndDevicesForUser(userId);
-                if (devs) {
-                    Object.keys(devs).forEach((deviceId) => {
-                        const d = DeviceInfo.fromStorage(devs[deviceId], deviceId);
-                        userStore[deviceId] = d;
-                    });
-                }
+                prom = prom.delay(5).then(() => {
```

Comment:
The delay is simply to try to spread out the work a bit and make the UI feel a bit snappier, by letting other things run in the event loop. It's an entirely subjective thing - it just made it feel a bit more responsive.

.all would make it run all the promises "in parallel" - or rather, in the same cycle of the event loop, which is exactly what we're trying to avoid here.